docs_nnx/guides/linen_to_nnx.rst

```diff
@@ -0,0 +1,732 @@
+
+But if you think closely, there actually wans't any ``jax.lax.scan`` operation at initialization time. What happened there is more like a ``jax.vmap`` operation - you are given a ``Block`` that accepts ``(in_dim, out_dim)``, and you "vmap" it over ``num_layers`` of times to create a larger array.
+
+This means that in NNX, since model initialization and running code are completely decoupled, we need to use ``nnx.vmap`` to initialize the underlying blocks, and then use ``nnx.scan`` to run the model input through them.
+
+For more information on NNX transforms, check out the `Transforms Guide <https://flax-nnx.readthedocs.build/en/guides/transforms.html>`__.
+
+.. codediff::
+  :title: Linen, NNX
+  :sync:
+
+  class Block(nn.Module):
+    features: int
+    training: bool
+
+    @nn.compact
+    def __call__(self, x, _):
+      x = nn.Dense(self.features)(x)
+      x = nn.Dropout(0.5)(x, deterministic=not self.training)
+      x = jax.nn.relu(x)
+      return x, None
+
+  class MLP(nn.Module):
+    features: int
+    num_layers: int
+
+
+
+
+    @nn.compact
+    def __call__(self, x, training: bool):
+      ScanBlock = nn.scan(
+        Block, variable_axes={'params': 0}, split_rngs={'params': True},
+        length=self.num_layers)
+
+      y, _ = ScanBlock(self.features, training)(x, None)
+      return y
+
+  model = MLP(64, num_layers=5)
+
+  ---
+
+  class Block(nnx.Module):
+    def __init__(self, input_dim, features, rngs):
+      self.linear = nnx.Linear(input_dim, features, rngs=rngs)
+      self.dropout = nnx.Dropout(0.5, rngs=rngs)
+
+    def __call__(self, x: jax.Array):  # No need to require a second input!
+      x = self.linear(x)
+      x = self.dropout(x)
+      x = jax.nn.relu(x)
+      return x   # No need to return a second output!
+
+  class MLP(nnx.Module):
+    def __init__(self, features, num_layers, rngs):
+      @nnx.split_rngs(splits=num_layers)
+      @nnx.vmap(in_axes=(0,), out_axes=0)
+      def create_block(rngs: nnx.Rngs):
+        return Block(features, features, rngs=rngs)
+
+      self.blocks = create_block(rngs)
+      self.num_layers = num_layers
+
+    def __call__(self, x):
+      @nnx.split_rngs(splits=self.num_layers)
+      @nnx.scan(in_axes=(nnx.Carry, 0), out_axes=nnx.Carry)
+      def forward(x, model):
+        x = model(x)
+        return x
+
+      return forward(x, self.blocks)
+
+  model = MLP(64, num_layers=5, rngs=nnx.Rngs(0))
+
+
+There are a few other details to explain in this example:
+
+* **What is that `nnx.split_rngs` decorator?** This is because ``jax.vmap`` and ``jax.lax.scan`` requires a list of RNG keys if each of its internal operations needs its own key. So for the 5 layers inside ``MLP``, it will split and provide 5 different RNG keys from its arguments before going into the JAX transform.
```

Comment:
Added explanation
```suggestion
* **What is that ``nnx.split_rngs`` decorator?** NNX transforms are completely agnostic of RNG state, this makes them behave more like JAX transforms but diverge from the Linen transforms which do handle RNG state. To regain this functionality, the ``nnx.split_rngs`` decorator allows you to split the ``Rngs`` before passing them
to the decorated function and 'lower' them afterwards so they can be used outside.
  * This is needed because ``jax.vmap`` and ``jax.lax.scan`` requires a list of RNG keys if each of its internal operations needs its own key. So for the 5 layers inside ``MLP``, it will split and provide 5 different RNG keys from its arguments before going into the JAX transform.
```